backend/windmill-api/src/openai.rs

```diff
@@ -71,63 +82,218 @@ lazy_static::lazy_static! {
     pub static ref OPENAI_AZURE_BASE_PATH: Option<String> = std::env::var("OPENAI_AZURE_BASE_PATH").ok();
 }
 
+#[derive(Deserialize)]
+struct OpenaiCredentials {
+    access_token: String,
+}
+async fn get_openai_key_using_credentials_flow(
+    mut resource: OpenaiClientCredentialsOauthResource,
+    db: &DB,
+    w_id: &String,
+) -> Result<String, Error> {
+    resource.client_id = get_variable_or_self(resource.client_id, &db, &w_id).await?;
+    resource.client_secret = get_variable_or_self(resource.client_secret, &db, &w_id).await?;
+    resource.token_url = get_variable_or_self(resource.token_url, &db, &w_id).await?;
+    let mut params = HashMap::new();
+    params.insert("grant_type", "client_credentials");
+    let response = HTTP_CLIENT
+        .post(resource.token_url)
+        .form(&params)
+        .basic_auth(resource.client_id, Some(resource.client_secret))
+        .send()
+        .await
+        .map_err(|err| {
+            Error::InternalErr(format!(
+                "Failed to get OpenAI credentials using credentials flow: {}",
+                err
+            ))
+        })?;
+    let response = response.json::<OpenaiCredentials>().await.map_err(|err| {
+        Error::InternalErr(format!(
+            "Failed to parse OpenAI credentials from credentials flow: {}",
+            err
+        ))
+    })?;
+    Ok(response.access_token)
+}
+
+struct OpenaiKeyCache {
+    api_key: String,
+    organization_id: Option<String>,
+    azure_base_path: Option<String>,
+    user: Option<String>,
+    expires_at: std::time::Instant,
+}
+
+impl OpenaiKeyCache {
+    fn new(
+        api_key: String,
+        organization_id: Option<String>,
+        azure_base_path: Option<String>,
+        expires_at: std::time::Instant,
+        user: Option<String>,
+    ) -> Self {
+        Self { api_key, organization_id, azure_base_path, expires_at, user }
+    }
+    fn is_expired(&self) -> bool {
+        self.expires_at < std::time::Instant::now()
+    }
+}
+
+lazy_static::lazy_static! {
+    static ref OPENAI_KEY_CACHE: Arc<RwLock<HashMap<String, OpenaiKeyCache>>> = Arc::new(RwLock::new(HashMap::new()));
+}
+
+#[derive(Deserialize)]
+struct ProxyQueryParams {
+    no_cache: Option<bool>,
+}
 async fn proxy(
     authed: ApiAuthed,
     Extension(db): Extension<DB>,
     Path((w_id, openai_path)): Path<(String, String)>,
-    body: Bytes,
+    Query(query_params): Query<ProxyQueryParams>,
+    mut body: Bytes,
 ) -> impl IntoResponse {
-    let mut tx = db.begin().await?;
-    let openai_resource_path = sqlx::query_scalar!(
-        "SELECT openai_resource_path FROM workspace_settings WHERE workspace_id = $1",
-        &w_id
-    )
-    .fetch_one(&mut *tx)
-    .await?;
-    tx.commit().await?;
+    let mut cache = OPENAI_KEY_CACHE.write().await;
```

Comment:
no that's terrible, you're taking a write lock on the full hashmap. Use quick_cache instead, see the cache of users or webhooks for instance